runtime/src/iree/hal/drivers/hip/pending_queue_actions.c

```diff
@@ -1064,11 +1312,147 @@ static void iree_hal_hip_worker_wait_pending_work_items(
+        (worker_state == IREE_HAL_HIP_WORKER_STATE_EXIT_REQUESTED);
+
+    iree_status_t status = iree_ok_status();
+    while (true) {
+      iree_hal_hip_atomic_slist_completion_t* entry =
+          iree_hal_hip_completion_slist_pop(worklist);
+      if (!entry) break;
+
+      const iree_hal_hip_dynamic_symbols_t* symbols = completion_area->symbols;
+      IREE_TRACE_ZONE_BEGIN_NAMED(z1, "hipEventSynchronize");
+      hipError_t result = symbols->hipEventSynchronize(entry->event);
+      IREE_TRACE_ZONE_END(z1);
+      if (IREE_UNLIKELY(result != hipSuccess)) {
+        // Let common destruction path take care of destroying the worklist.
+        // When we know all host stream callbacks are done and not touching
+        // anything.
+        iree_hal_hip_completion_slist_push(worklist, entry);
+        break;
+      }
+      entry->callback(entry->user_data);
+
+      if (IREE_UNLIKELY(entry->created_event)) {
+        IREE_HIP_IGNORE_ERROR(symbols, hipEventDestroy(entry->event));
+      }
+      iree_allocator_free(completion_area->host_allocator, entry);
+
+      // Now we fully executed and cleaned up this entry. Decrease the work
+      // items counter.
+      iree_hal_hip_queue_decrement_completion_count(completion_area);
+    }
+
+    if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+      iree_hal_hip_completion_wait_pending_completion_items(completion_area);
+      iree_hal_hip_post_error_to_completion_state(completion_area,
+                                                  iree_status_code(status));
+      IREE_TRACE_ZONE_END(z0);
+      return -1;
+    }
+
+    if (IREE_UNLIKELY(should_exit &&
+                      iree_hal_hip_completion_has_no_pending_completion_items(
+                          completion_area))) {
+      iree_hal_hip_completion_has_no_pending_completion_items(completion_area);
+      // Signal that this thread is committed to exit.
+      // This state has a priority that is only lower than error exit.
+      // A HIP callback may have posted an error, make sure we don't
+      // overwrite this error state.
+      iree_hal_hip_worker_state_t prev_state =
+          IREE_HAL_HIP_WORKER_STATE_EXIT_REQUESTED;
+      iree_atomic_compare_exchange_strong_int32(
+          &completion_area->worker_state, /*expected=*/&prev_state,
+          /*desired=*/IREE_HAL_HIP_WORKER_STATE_EXIT_COMMITTED,
+          /*order_succ=*/iree_memory_order_acq_rel,
+          /*order_fail=*/iree_memory_order_acquire);
+      iree_notification_post(&completion_area->exit_notification,
+                             IREE_ALL_WAITERS);
+      IREE_TRACE_ZONE_END(z0);
+      return 0;
+    }
+    IREE_TRACE_ZONE_END(z0);
+  }
+
+  return 0;
+}
+
 // The main function for the ready-list processing worker thread.
 static int iree_hal_hip_worker_execute(
     iree_hal_hip_working_area_t* working_area) {
   iree_hal_hip_ready_action_slist_t* worklist = &working_area->ready_worklist;
 
+  iree_status_t status = IREE_HIP_RESULT_TO_STATUS(
```

Comment:
Would be good to put some comments around this as a general reminder for folks reading the code, given it bite us once :)